Encountering new ClassNotFoundException when attempting to inflate certain layouts

I'm now encountering a startup crash when running a handful of projects on device. This error is not reproducible on 15.9 or earlier dev16 previews.

### Steps to Reproduce

1. Download and run the following project:
https://github.com/xamarin/monodroid-samples/tree/master/android-o/EmojiCompat


<!--
If you have a repro project, you may drag & drop the .zip/etc. onto the issue editor to attach it.
-->

### Expected Behavior
Layout XML files which specify a fully qualified class name to a layout in a .cs file should inflate successfully.

### Actual Behavior
```
12-07 10:27:07.762 16463 16463 E AndroidRuntime: Caused by: android.view.InflateException: Binary XML file line #1: Binary XML file line #1: Error inflating class EmojiCompatSample.CustomTextView
12-07 10:27:07.762 16463 16463 E AndroidRuntime: Caused by: android.view.InflateException: Binary XML file line #1: Error inflating class EmojiCompatSample.CustomTextView
12-07 10:27:07.762 16463 16463 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "EmojiCompatSample.CustomTextView" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/com.xamarin.EmojiCompat-e5BLLTVnZZzpdFtPjSKlXA==/base.apk"],nativeLibraryDirectories=[/data/app/com.xamarin.EmojiCompat-e5BLLTVnZZzpdFtPjSKlXA==/lib/arm64, /system/fake-libs64, /data/app/com.xamarin.EmojiCompat-e5BLLTVnZZzpdFtPjSKlXA==/base.apk!/lib/arm64-v8a, /system/lib64]]
12-07 10:27:07.762 16463 16463 E AndroidRuntime:        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
```

### Version Information
Microsoft Visual Studio Enterprise 2019 [28406.54.d16.0]
Xamarin.Android SDK   9.1.199.32 (HEAD/2e6d2a8d)
    Mono: mono/mono/2018-06@b63e5378e38
    Java.Interop: xamarin/java.interop/master@24e7725
    LibZipSharp: grendello/LibZipSharp/master@de1712c
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
    MXE: xamarin/mxe/xamarin@b9cbb535
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.25.2@4ea4c69
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-0-p2@9e78d6e
<!--
1. On macOS and within Visual Studio, select Visual Studio > About Visual Studio, then click the Show Details button, then click the Copy Information button.

2. Paste below this comment block.
-->

### Log File
logcat: 
https://gist.github.com/pjcollins/06f2177156609585b506f406522f5a8c

build: 
https://gist.githubusercontent.com/pjcollins/770ac072099e9eb70b69b52b92387ac8/raw/dbcc8af3d28660dd716f062138f97186e8335780/gistfile1.txt
<!--
1. On macOS and within Visual Studio:
    a. Click **Tools** > **SDK Command Prompt**.
    b. Within the launched `Terminal.app` window, run:

            adb logcat -d | pbcopy

2. Paste below this comment block
-->



<!--
Switch to the "Preview" tab to ensure your issue renders correctly.

Please add an appropriate "Area:" label in the Labels section.
-->

--- a/src/Xamarin.Android.Build.Tasks/Tasks/ConvertCustomView.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tasks/ConvertCustomView.cs
@@ -84,8 +84,8 @@ namespace Xamarin.Android.Tasks {
 				if (resdir == null) {
 					continue;
 				}
-				var hash = resdir?.GetMetadata ("Hash") ?? null;
-				var stamp = resdir?.GetMetadata ("StampFile") ?? null;
+				var hash = resdir.GetMetadata ("Hash");
+				var stamp = resdir.GetMetadata ("StampFile");
 				var filename = !string.IsNullOrEmpty (hash) ? hash : "compiled";
 				var stampFile = !string.IsNullOrEmpty (stamp) ? stamp : $"{filename}.stamp";
 				Log.LogDebugMessage ($"{filename} {stampFile}");

--- a/src/Xamarin.Android.Build.Tasks/Tasks/ConvertCustomView.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tasks/ConvertCustomView.cs
@@ -23,6 +23,9 @@ namespace Xamarin.Android.Tasks {
 
 		public ITaskItem [] ResourceDirectories { get; set; }
 
+		[Output]
+		public ITaskItem [] Processed { get; set; }
+
 		public override bool Execute ()
 		{
 			var resource_name_case_map = MonoAndroidHelper.LoadResourceCaseMap (ResourceNameCaseMap);
@@ -75,7 +78,20 @@ namespace Xamarin.Android.Tasks {
 					}
 				}
 			}
-
+			var output = new List<ITaskItem> ();
+			foreach (var file in processed) {
+				ITaskItem resdir = ResourceDirectories?.FirstOrDefault (x => file.StartsWith (x.ItemSpec)) ?? null;
+				var hash = resdir?.GetMetadata ("Hash") ?? null;
+				var stamp = resdir?.GetMetadata ("StampFile") ?? null;
+				var filename = !string.IsNullOrEmpty (hash) ? hash : "compiled";
+				var stampFile = !string.IsNullOrEmpty (stamp) ? stamp : $"{filename}.stamp";
+				Log.LogDebugMessage ($"{filename} {stampFile}");
+				output.Add (new TaskItem (file, new Dictionary<string, string> {
+					{ "StampFile" , $"{stampFile}" },
+					{ "Hash" , $"{filename}" },
+				}));
+			}
+			Processed = output.ToArray ();
 			return !Log.HasLoggedErrors;
 		}
 

--- a/src/Xamarin.Android.Build.Tasks/Tasks/ConvertCustomView.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tasks/ConvertCustomView.cs
@@ -81,6 +81,9 @@ namespace Xamarin.Android.Tasks {
 			var output = new List<ITaskItem> ();
 			foreach (var file in processed) {
 				ITaskItem resdir = ResourceDirectories?.FirstOrDefault (x => file.StartsWith (x.ItemSpec)) ?? null;
+				if (resdir == null) {
+					continue;
+				}
 				var hash = resdir?.GetMetadata ("Hash") ?? null;
 				var stamp = resdir?.GetMetadata ("StampFile") ?? null;
 				var filename = !string.IsNullOrEmpty (hash) ? hash : "compiled";

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AndroidUpdateResourcesTest.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AndroidUpdateResourcesTest.cs
@@ -461,6 +461,8 @@ namespace UnnamedProject
 				Assert.AreEqual ("@color/deep_purple_A200", item.Value, "item value should be @color/deep_purple_A200");
 				Assert.IsFalse (StringAssertEx.ContainsText (b.LastBuildOutput, "AndroidResgen: Warning while updating Resource XML"),
 					"Warning while processing resources should not have been raised.");
+				Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
+				Assert.IsTrue (b.IsTargetSkipped ("_GenerateJavaStubs"), "Target _GenerateJavaStubs should have been skipped");
 				Assert.IsTrue (b.Clean (proj), "Clean should have succeeded.");
 			}
 		}

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AndroidUpdateResourcesTest.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AndroidUpdateResourcesTest.cs
@@ -462,7 +462,7 @@ namespace UnnamedProject
 				Assert.IsFalse (StringAssertEx.ContainsText (b.LastBuildOutput, "AndroidResgen: Warning while updating Resource XML"),
 					"Warning while processing resources should not have been raised.");
 				Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
-				Assert.IsTrue (b.IsTargetSkipped ("_GenerateJavaStubs"), "Target _GenerateJavaStubs should have been skipped");
+				Assert.IsTrue (b.Output.IsTargetSkipped ("_GenerateJavaStubs"), "Target _GenerateJavaStubs should have been skipped");
 				Assert.IsTrue (b.Clean (proj), "Clean should have succeeded.");
 			}
 		}

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AndroidUpdateResourcesTest.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AndroidUpdateResourcesTest.cs
@@ -310,16 +310,11 @@ using System.Runtime.CompilerServices;
 		/// <summary>
 		/// Based on https://bugzilla.xamarin.com/show_bug.cgi?id=29263
 		/// </summary>
-		public void CheckXmlResourcesFilesAreProcessed ([Values(false, true)] bool isRelease)
+		public void CheckXmlResourcesFilesAreProcessed ([Values(false, true)] bool useAapt2)
 		{
-			var projectPath = String.Format ("temp/CheckXmlResourcesFilesAreProcessed_{0}", isRelease);
+			var projectPath = String.Format ("temp/CheckXmlResourcesFilesAreProcessed_{0}", useAapt2);
 
-			var lib = new XamarinAndroidLibraryProject () {
-				IsRelease = isRelease,
-				ProjectName = "Classlibrary1",
-			};
-			lib.AndroidResources.Add (new AndroidItem.AndroidResource ("Resources\\layout\\custom_text.xml") {
-				TextContent = () => @"<?xml version=""1.0"" encoding=""utf-8"" ?>
+			var layout =  @"<?xml version=""1.0"" encoding=""utf-8"" ?>
 <LinearLayout xmlns:android=""http://schemas.android.com/apk/res/android""
 	android:orientation = ""vertical""
 	android:layout_width = ""fill_parent""
@@ -334,7 +329,12 @@ using System.Runtime.CompilerServices;
 		android:layout_width = ""fill_parent""
 		android:layout_height = ""wrap_content""
 		android:text = ""namespace_proper"" />
-</LinearLayout>"
+</LinearLayout>";
+			var lib = new XamarinAndroidLibraryProject () {
+				ProjectName = "Classlibrary1",
+			};
+			lib.AndroidResources.Add (new AndroidItem.AndroidResource ("Resources\\layout\\custom_text_lib.xml") {
+				TextContent = () => layout,
 			});
 			lib.Sources.Add (new BuildItem.Source ("CustomTextView.cs") {
 				TextContent = () => @"using Android.Widget;
@@ -352,12 +352,17 @@ namespace ClassLibrary1
 			});
 
 			var proj = new XamarinAndroidApplicationProject () {
-				IsRelease = isRelease,
 				OtherBuildItems = {
 					new BuildItem.ProjectReference (@"..\Classlibrary1\Classlibrary1.csproj", "Classlibrary1", lib.ProjectGuid) {
 					},
 				}
 			};
+
+			proj.SetProperty ("AndroidUseAapt2", useAapt2.ToString ());
+
+			proj.AndroidResources.Add (new AndroidItem.AndroidResource ("Resources\\layout\\custom_text_app.xml") {
+				TextContent = () => layout,
+			});
 
 			proj.AndroidResources.Add (new AndroidItem.AndroidResource ("Resources\\drawable\\UPPER_image.png") {
 				BinaryContent = () => XamarinAndroidCommonProject.icon_binary_mdpi
@@ -421,25 +426,21 @@ namespace UnnamedProject
 			proj.PackageReferences.Add (KnownPackages.SupportV7AppCompat_22_1_1_1);
 			proj.PackageReferences.Add (KnownPackages.SupportV7Palette_22_1_1_1);
 			proj.SetProperty ("TargetFrameworkVersion", "v5.0");
-			proj.SetProperty (proj.DebugProperties, "JavaMaximumHeapSize", "1G");
-			proj.SetProperty (proj.ReleaseProperties, "JavaMaximumHeapSize", "1G");
-			using (var libb = CreateDllBuilder (Path.Combine (projectPath, lib.ProjectName)))
-			using (var b = CreateApkBuilder (Path.Combine (projectPath, proj.ProjectName))) {
-				b.Verbosity = LoggerVerbosity.Diagnostic;
+			using (var libb = CreateDllBuilder (Path.Combine (projectPath, lib.ProjectName), cleanupOnDispose: false))
+			using (var b = CreateApkBuilder (Path.Combine (projectPath, proj.ProjectName), cleanupOnDispose: false)) {
 				Assert.IsTrue (libb.Build (lib), "Library Build should have succeeded.");
 				Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
-				var customViewPath = Path.Combine (Root, b.ProjectDirectory, proj.IntermediateOutputPath, "lp", "0", "jl", "res", "layout", "custom_text.xml");
-				Assert.IsTrue (File.Exists (customViewPath), $"custom_text.xml should exist at {customViewPath}");
-				var doc = XDocument.Load (customViewPath);
-				Assert.IsNotNull (doc.Element ("LinearLayout"), "PreferenceScreen should be present in preferences.xml");
-				Assert.IsNull (doc.Element ("LinearLayout").Element ("Classlibrary1.CustomTextView"),
-					"Classlibrary1.CustomTextView should have been replaced with an $(MD5Hash).CustomTextView");
-				Assert.IsNull (doc.Element ("LinearLayout").Element ("classlibrary1.CustomTextView"),
-					"classlibrary1.CustomTextView should have been replaced with an $(MD5Hash).CustomTextView");
+				var intermediate = Path.Combine (Root, b.ProjectDirectory, proj.IntermediateOutputPath);
+				var packaged_resources = Path.Combine (intermediate, "android", "bin", "packaged_resources");
+				FileAssert.Exists (packaged_resources);
+				using (var zip = ZipHelper.OpenZip (packaged_resources)) {
+					CheckCustomView (zip, intermediate, "lp", "0", "jl", "res", "layout", "custom_text_lib.xml");
+					CheckCustomView (zip, intermediate, "res", "layout", "custom_text_app.xml");
+				}
 				
 				var preferencesPath = Path.Combine (Root, b.ProjectDirectory, proj.IntermediateOutputPath, "res","xml","preferences.xml");
 				Assert.IsTrue (File.Exists (preferencesPath), "Preferences.xml should have been renamed to preferences.xml");
-				doc = XDocument.Load (preferencesPath);
+				var doc = XDocument.Load (preferencesPath);
 				Assert.IsNotNull (doc.Element ("PreferenceScreen"), "PreferenceScreen should be present in preferences.xml");
 				Assert.IsNull (doc.Element ("PreferenceScreen").Element ("UnnamedProject.CustomPreference"),
 					"UnamedProject.CustomPreference should have been replaced with an $(MD5Hash).CustomPreference");
@@ -458,9 +459,41 @@ namespace UnnamedProject
 					.FirstOrDefault (x => x.Attribute("name").Value == "colorAccent");
 				Assert.IsNotNull (item, "The Style should contain an Item");
 				Assert.AreEqual ("@color/deep_purple_A200", item.Value, "item value should be @color/deep_purple_A200");
-				StringAssertEx.DoesNotContain ("AndroidResgen: Warning while updating Resource XML", b.LastBuildOutput,
+				Assert.IsFalse (StringAssertEx.ContainsText (b.LastBuildOutput, "AndroidResgen: Warning while updating Resource XML"),
 					"Warning while processing resources should not have been raised.");
 				Assert.IsTrue (b.Clean (proj), "Clean should have succeeded.");
+			}
+		}
+
+		void CheckCustomView (Xamarin.Tools.Zip.ZipArchive zip, params string [] paths)
+		{
+			var customViewPath = Path.Combine (paths);
+			FileAssert.Exists (customViewPath, $"custom_text.xml should exist at {customViewPath}");
+			var doc = XDocument.Load (customViewPath);
+			Assert.IsNotNull (doc.Element ("LinearLayout"), "PreferenceScreen should be present in preferences.xml");
+			Assert.IsNull (doc.Element ("LinearLayout").Element ("Classlibrary1.CustomTextView"),
+				"Classlibrary1.CustomTextView should have been replaced with an $(MD5Hash).CustomTextView");
+			Assert.IsNull (doc.Element ("LinearLayout").Element ("classlibrary1.CustomTextView"),
+				"classlibrary1.CustomTextView should have been replaced with an $(MD5Hash).CustomTextView");
+
+			//Now check the zip
+			var customViewInZip = "res/layout/" + Path.GetFileName (customViewPath);
+			var entry = zip.ReadEntry (customViewInZip);
+			Assert.IsNotNull (entry, $"`{customViewInZip}` should exist in packaged_resources!");
+
+			using (var stream = new MemoryStream ()) {
+				entry.Extract (stream);
+				stream.Position = 0;
+
+				using (var reader = new StreamReader (stream)) {
+					//NOTE: This is a binary format, but we can still look for text within.
+					//      Don't use `StringAssert` because `contents` make the failure message unreadable.
+					var contents = reader.ReadToEnd ();
+					Assert.IsFalse (contents.Contains ("Classlibrary1.CustomTextView"),
+							"Classlibrary1.CustomTextView should have been replaced with an $(MD5Hash).CustomTextView");
+					Assert.IsFalse (contents.Contains ("classlibrary1.CustomTextView"),
+							"classlibrary1.CustomTextView should have been replaced with an $(MD5Hash).CustomTextView");
+				}
 			}
 		}
 

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ZipArchiveExTests.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ZipArchiveExTests.cs
@@ -48,7 +48,7 @@ namespace Xamarin.Android.Build.Tests
 				}
 			}
 
-			Assert.AreEqual (expected.Trim (), builder.ToString ().Trim ());
+			Assert.AreEqual (expected.Trim ().Replace("\r\n", "\n"), builder.ToString ().Trim ().Replace("\r\n", "\n"));
 		}
 
 		[Test]

--- a/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets
+++ b/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets
@@ -1423,13 +1423,17 @@ because xbuild doesn't support framework reference assemblies.
 		ToolPath="$(Aapt2ToolPath)"
 		ToolExe="$(Aapt2ToolExe)"
 	/>
+	<ItemGroup>
+		<_HashStampFiles Include="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')" />
+		<_HashFlataFiles Include="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).flata')" />
+	</ItemGroup>
 	<Touch
-			Files="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')"
+			Files="@(_HashStampFiles)"
 			AlwaysCreate="True"
 	/>
 	<ItemGroup>
-		<FileWrites Include="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).flata')" />
-		<FileWrites Include="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')" />
+		<FileWrites Include="@(_HashStampFiles)" />
+		<FileWrites Include="@(_HashFlataFiles)" />
 	</ItemGroup>
 </Target>
 

--- a/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets
+++ b/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets
@@ -1409,8 +1409,8 @@ because xbuild doesn't support framework reference assemblies.
 
 <Target Name="_CompileAndroidLibraryResources" DependsOnTargets="_ConvertLibraryResourcesCases"
 		Condition="'$(_AndroidUseAapt2)' == 'True'"
-		Inputs="%(_LibraryResourceHashDirectories.StampFile)"
-		Outputs="$(_AndroidLibraryFlatArchivesDirectory)%(_LibraryResourceHashDirectories.Hash).stamp"
+		Inputs="@(_LibraryResourceHashDirectories->'%(StampFile)')"
+		Outputs="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')"
 	>
 	<MakeDir Directories="$(_AndroidLibraryFlatArchivesDirectory)" Condition="!Exists('$(_AndroidLibraryFlatArchivesDirectory)')" />
 	<Aapt2Compile
@@ -1424,19 +1424,19 @@ because xbuild doesn't support framework reference assemblies.
 		ToolExe="$(Aapt2ToolExe)"
 	/>
 	<Touch
-			Files="$(_AndroidLibraryFlatArchivesDirectory)%(_LibraryResourceHashDirectories.Hash).stamp"
+			Files="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')"
 			AlwaysCreate="True"
 	/>
 	<ItemGroup>
-		<FileWrites Include="$(_AndroidLibraryFlatArchivesDirectory)%(_LibraryResourceHashDirectories.Hash).flata" />
-		<FileWrites Include="$(_AndroidLibraryFlatArchivesDirectory)%(_LibraryResourceHashDirectories.Hash).stamp" />
+		<FileWrites Include="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).flata')" />
+		<FileWrites Include="@(_LibraryResourceHashDirectories->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')" />
 	</ItemGroup>
 </Target>
 
 <Target Name="_CompileResources"
 		Condition="'$(_AndroidUseAapt2)' == 'True'"
 		Inputs="@(AndroidResource)"
-		Outputs="$(_AndroidLibraryFlatArchivesDirectory)\compiled.flata"
+		Outputs="$(_AndroidLibraryFlatArchivesDirectory)\compiled.stamp"
 	>
 	<MakeDir Directories="$(_AndroidLibraryFlatArchivesDirectory)" Condition="!Exists('$(_AndroidLibraryFlatArchivesDirectory)')" />
 	<!-- Change cases so we support mixed case resource names -->
@@ -1459,8 +1459,10 @@ because xbuild doesn't support framework reference assemblies.
 		ToolPath="$(Aapt2ToolPath)"
 		ToolExe="$(Aapt2ToolExe)"
 	/>
+	<Touch Files="$(_AndroidLibraryFlatArchivesDirectory)\compiled.stamp" AlwaysCreate="True" />
 	<ItemGroup>
 		<FileWrites Include="$(_AndroidLibraryFlatArchivesDirectory)\compiled.flata" />
+		<FileWrites Include="$(_AndroidLibraryFlatArchivesDirectory)\compiled.stamp" />
 	</ItemGroup>
 </Target>
 
@@ -2350,8 +2352,28 @@ because xbuild doesn't support framework reference assemblies.
 	Condition="Exists('$(_CustomViewMapFile)')"
 	CustomViewMapFile="$(_CustomViewMapFile)"
 	AcwMapFile="$(_AcwMapFile)"
-	ResourceDirectories="$(MonoAndroidResDirIntermediate);@(LibraryResourceDirectories)"
+	ResourceDirectories="$(MonoAndroidResDirIntermediate);@(_LibraryResourceHashDirectories)"
 	ResourceNameCaseMap="$(_AndroidResourceNameCaseMap)"
+  >
+    <Output TaskParameter="Processed" ItemName="_ProcessedCustomViews" />	
+  </ConvertCustomView>
+  <Delete Files="@(_ProcessedCustomViews->'$(_AndroidLibraryFlatArchivesDirectory)%(Hash).stamp')"
+    Condition=" '$(AndroidUseAapt2)' == 'True' And '@(_ProcessedCustomViews)' != '' "
+  />
+  <Aapt2Compile
+    Condition=" '$(AndroidUseAapt2)' == 'True' And '@(_ProcessedCustomViews)' != '' "
+    ContinueOnError="$(DesignTimeBuild)"
+    ResourceDirectories="@(_LibraryResourceHashDirectories);$(MonoAndroidResDirIntermediate)"
+    ExplicitCrunch="$(AndroidExplicitCrunch)"
+    ExtraArgs="$(AndroidAapt2CompileExtraArgs)"
+    FlatArchivesDirectory="$(_AndroidLibraryFlatArchivesDirectory)"
+    ToolPath="$(Aapt2ToolPath)"
+    ToolExe="$(Aapt2ToolExe)">
+    <Output TaskParameter="CompiledResourceFlatArchives" ItemName="_UpdatedFlatArchives" />	
+  </Aapt2Compile>
+  <Touch Files="@(_UpdatedFlatArchives->'$(_AndroidLibraryFlatArchivesDirectory)\%(Filename).stamp')"
+    Condition=" '$(AndroidUseAapt2)' == 'True' And '@(_UpdatedFlatArchives)' != '' "
+    AlwaysCreate="True"
   />
   <Touch Files="$(_AndroidStampDirectory)_GenerateJavaStubs.stamp" AlwaysCreate="True" />
 </Target>

